### PR TITLE
Feature/sasrec

### PR DESCRIPTION
**Pull request by blondered**

### CODE REVIEW

File: rectools/models/sasrec.py [hunk 1]
```diff
@@ -0,0 +1,723 @@
...
+from rectools import Columns, ExternalIds
+from rectools.dataset import Dataset, Interactions
+from rectools.dataset.identifiers import IdMap
+from rectools.models.base import ErrorBehaviour, InternalRecoTriplet, ModelBase
+from rectools.models.rank import Distance, ImplicitRanker
+from rectools.types import InternalIdsArray
+
+PADDING_VALUE = "PAD"
+
+logger = logging.getLogger(__name__)
```
Comment (feldlime) [inline rectools/models/sasrec.py:24]: It's nice to have logging, but we need a proper setup.  
Now it's used only in one place and quite useless. Let's remove it for now

File: rectools/models/sasrec.py [hunk 2]
```diff
@@ -0,0 +1,723 @@
...
+            for epoch in range(epoch_start_idx, self.epochs + 1):
+                logger.info("training epoch %s", epoch)
+                for x, y, w in fit_dataloader:
+                    x = x.to(self.device)  # [batch_size, session_maxlen]
+                    y = y.to(self.device)  # [batch_size, session_maxlen]
+                    w = w.to(self.device)  # [batch_size, session_maxlen]
+
+                    self.train_step(x, y, w)
+
+        except KeyboardInterrupt:
```
Comment (feldlime) [inline rectools/models/sasrec.py:343]: We don't really need this try..except block

File: rectools/models/sasrec.py [hunk 3]
```diff
@@ -0,0 +1,723 @@
...
+        self.model = self.model.eval()
+        self.model.to(self.device)
+
+        # Dataset has already been filtered and adapted to known item_id_map
+        recommend_dataloader = self.data_preparator.get_dataloader_recommend(dataset)
+
+        session_embs = []
+        item_embs = self.model.item_model.get_all_embeddings()  # [n_items + 1, factors]
+        with torch.no_grad():
+            for x_batch in tqdm.tqdm(recommend_dataloader):
```
Comment (feldlime) [inline rectools/models/sasrec.py:664]: It's probably better to use `from tqdm.auto import tqdm` for pretty bars in jupyter. But it's not quite safe. Please add TODO here for the future

File: rectools/models/sasrec.py [hunk 4]
```diff
@@ -0,0 +1,723 @@
...
+        raise NotImplementedError()
+
+
+class IdEmbeddingsItemNet(ItemNetBase):
+    """
+    Base class for item embeddings. To use more complicated logic then just id embeddings inherit
+    from this class and pass your custom ItemNet to your model params
+    """
+
+    def __init__(self, factors: int, n_items: int, dropout_rate: float):
```
[the remark this answers is not in the document]
Comment (blondered) [reply rectools/models/sasrec.py:62]: I like `n_factors` more. But we have `factors` in dssm(

File: rectools/models/sasrec.py (hunk 2 above)
Comment (blondered) [reply rectools/models/sasrec.py:343]: Agreed. Trainer is being completely refactored right now. We didn't do it yet. And we are also moving to Lightning module and trainer in one of the nearest PRs.

File: rectools/models/sasrec.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (blondered) [reply rectools/models/sasrec.py:62]: Actually no. Not in dssm. We have `factors` in implicit ALS. 
We can move to `n_factors`, not too bad

[the remark this answers is not in the document]
Comment (blondered) [reply rectools/models/sasrec.py:62]: Another update. I found it. We have `factors` in `PureSVDModel`

[the remark this answers is not in the document]
Comment (feldlime) [reply rectools/models/sasrec.py:62]: - in ALS it's an inner implicit thing, that's fine
- we can support both in PureSVD for consistency
-

File: rectools/models/sasrec.py (hunk 1 above)
Comment (feldlime) [reply rectools/models/sasrec.py:24]: Ok, then please just add TODO

File: rectools/models/sasrec.py (hunk 2 above)
Comment (feldlime) [reply rectools/models/sasrec.py:343]: ok, just add TODO please

File: rectools/models/sasrec.py [hunk 5]
```diff
@@ -0,0 +1,723 @@
...
+
+class ItemNetBase(nn.Module):
+    """Base class ItemNet. Used only for type hinting."""
+
+    def forward(self, items: torch.Tensor) -> torch.Tensor:
+        """TODO"""
+        raise NotImplementedError()
+
+    @classmethod
+    def from_dataset(cls: tp.Type[ItemNetT], dataset: Dataset, *args: tp.Any, **kwargs: tp.Any) -> ItemNetT:
```
[the remark this answers is not in the document]
Comment (blondered) [reply rectools/models/sasrec.py]: fixed

File: rectools/models/sasrec.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (blondered) [reply rectools/models/sasrec.py]: changed to `n_factors`

File: rectools/models/sasrec.py (hunk 1 above)
Comment (blondered) [reply rectools/models/sasrec.py]: fixed

File: rectools/models/sasrec.py (hunk 2 above)
Comment (blondered) [reply rectools/models/sasrec.py]: removed it

File: rectools/models/sasrec.py (hunk 3 above)
Comment (blondered) [reply rectools/models/sasrec.py]: fixed

**Review by feldlime — APPROVED** [review]

**Pull request merged**